src/Castle.Windsor/MicroKernel/Arguments.cs

```diff
@@ -17,210 +17,239 @@ namespace Castle.MicroKernel
-			}
-			else
-			{
-				arguments = new Dictionary<object, object>(new ArgumentsComparerExtended(customComparers));
+				EnsureWritable();
+				dictionary[key] = value;
 			}
 		}
 
-		public int Count
-		{
-			get { return arguments.Count; }
-		}
+		public int Count => dictionary.Count;
 
-		public object this[object key]
-		{
-			get { return arguments[key]; }
-			set { arguments[key] = value; }
-		}
+		public ICollection Keys => dictionary.Keys;
 
-		public ICollection Keys
-		{
-			get { return arguments.Keys; }
-		}
+		public ICollection Values => dictionary.Values;
 
-		public ICollection Values
-		{
-			get { return arguments.Values; }
-		}
+		bool ICollection.IsSynchronized => dictionary.IsSynchronized;
 
-		bool ICollection.IsSynchronized
-		{
-			get { return arguments.IsSynchronized; }
-		}
+		object ICollection.SyncRoot => dictionary.SyncRoot;
 
-		object ICollection.SyncRoot
-		{
-			get { return arguments.SyncRoot; }
-		}
+		bool IDictionary.IsFixedSize => dictionary.IsFixedSize;
 
-		bool IDictionary.IsFixedSize
-		{
-			get { return arguments.IsFixedSize; }
-		}
+		bool IDictionary.IsReadOnly => isReadOnly;
 
-		bool IDictionary.IsReadOnly
-		{
-			get { return arguments.IsReadOnly; }
-		}
+		public static readonly Arguments Empty = new Arguments() { isReadOnly = true };
 
 		public void Add(object key, object value)
 		{
-			arguments.Add(key, value);
+			EnsureWritable();
+			dictionary.Add(key, value);
 		}
 
 		public void Clear()
 		{
-			arguments.Clear();
+			EnsureWritable();
+			dictionary.Clear();
 		}
 
 		public bool Contains(object key)
 		{
-			return arguments.Contains(key);
+			return dictionary.Contains(key);
 		}
 
 		public IDictionaryEnumerator GetEnumerator()
 		{
-			return arguments.GetEnumerator();
+			return dictionary.GetEnumerator();
 		}
 
 		public void Remove(object key)
 		{
-			arguments.Remove(key);
+			EnsureWritable();
+			dictionary.Remove(key);
 		}
 
 		public Arguments Clone()
 		{
-			var dictionary = arguments as Dictionary<object, object>;
-			if (dictionary != null)
+			if (dictionary is Dictionary<object, object>)
 			{
-				var comparerExtended = dictionary.Comparer as ArgumentsComparerExtended;
-				if (comparerExtended != null)
-				{
-					return new Arguments(arguments, comparerExtended.CustomComparers);
-				}
+				return new Arguments(dictionary);
 			}
 
-			return new Arguments(arguments);
+			return new Arguments(dictionary);
 		}
 
 		void ICollection.CopyTo(Array array, int index)
 		{
-			arguments.CopyTo(array, index);
+			dictionary.CopyTo(array, index);
 		}
 
 		IEnumerator IEnumerable.GetEnumerator()
 		{
 			return GetEnumerator();
 		}
 
-		private class ArgumentsComparer : IEqualityComparer<object>
+		/// <summary>
+		/// Inserts a new named argument with given key. If an argument for this name already exists, it will be overwritten.
+		/// </summary>
+		public Arguments Insert(string key, object value)
```

Comment:
I also think we should mark the methods as virtual, so folks out there can inherit and override our API without having to raise PR's for further type support.